### PR TITLE
train_model(): evaluates on test set the best model (not the last); returns the best model (not the last)

### LINKED ISSUE #1192: `ModelTestCase.ensure_model_can_train_save_and_load` compares last-epoch weights with serialized *best* weights
in particular, if your test case model trained (say) 10 epochs, and got the best results in epoch 5, then the test would fail because it would compare `model.state_dict()` (i.e. epoch 10) with `archive.model.state_dict()` (the weights from epoch 5)

in the meantime I am setting `num_epoch = 1` as a workaround

### PR DESCRIPTION
**Pull request by janLuke**
Quick fix of the bug #1223: setting "evaluate_on_test", the **last** model was evaluated, not the **best** one (as one expected). 

Furthermore, now train_model() returns the best model rather than the last, which fixes #1192.

### CODE REVIEW

File: allennlp/commands/train.py [hunk 1]
```diff
@@ -300,8 +306,15 @@ def train_model(params: Params,
     # Now tar up results
     archive_model(serialization_dir, files_to_archive=params.files_to_archive)
 
+    logger.info("Loading the best epoch weights.")
+    best_model_state_path = os.path.join(serialization_dir, 'best.th')
```
Comment (nelson-liu) [inline allennlp/commands/train.py]: instead of hardcoding `best.th`, maybe import `_DEFAULT_WEIGHTS` from model? https://github.com/allenai/allennlp/blob/master/allennlp/models/model.py#L25

Comment (janLuke) [reply allennlp/commands/train.py]: Good point. I actually looked for a constant with the strict semantic of "best weights file name" before hard-coding it. I saw ``_DEFAULT_WEIGHTS`` but: 1) I think it has another semantic and 2) it's "declared" as "private". Furthermore, ``"best.th"`` is hard-coded elsewhere, so I hard-coded it. I think we could introduce a constant ``BEST_WEIGHTS`` or better (though longer) ``BEST_WEIGHTS_FILENAME``.

Comment (janLuke) [reply allennlp/commands/train.py]: Actually, looking again at model.py, I think that ``_DEFAULT_WEIGHTS`` could be safely renamed as ``BEST_WEIGHTS`` or ``BEST_WEIGHTS_FILENAME``.

**Comment by nelson-liu** [comment]
hmm, I think https://github.com/allenai/allennlp/pull/1225/commits/ea01031e3cb026620c40f69320b78c514e0a8c0a should go into another PR?

**Comment by janLuke** [comment]
Damn, sorry, I did this by mistake ^.^". Not a git expert here. How can I fix it?

**Comment by nelson-liu** [comment]
hmm, it's a bit tough since you're working on master--I'd recommend starting [a new branch](https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell) when implementing a feature in the future. do you mind if i cherry pick your commits and merge them from there?

**Comment by janLuke** [comment]
@nelson-liu I created a new PR from a new branch with a single commit (I cherry picked the 2nd commit of this PR and wrote a better description): https://github.com/allenai/allennlp/pull/1234. I'm closing this one. 

Hopefully I didn't make mistakes with the new one. Have patience, I've not much experience with git(hub) ^.^

@DeNeutoy Happy to lend a (small) hand :)
